crates/stress/src/bin/sim2h_stress.rs

```diff
@@ -0,0 +1,357 @@
+            },
+        )));
+    }
+
+    /// send a ping message to sim2h
+    pub fn ping(&mut self) {
+        self.send_wire(WireMessage::Ping);
+    }
+}
+
+impl StressJob for Job {
+    /// check for any messages from sim2h and also send a ping
+    fn tick(&mut self, logger: &mut StressJobMetricLogger) -> StressJobTickResult {
+        let (_, evs) = self.stream_manager.process().unwrap();
+        for ev in evs {
+            match ev {
+                StreamEvent::ErrorOccured(_, e) => panic!("{:?}", e),
+                StreamEvent::ConnectResult(_, _) => panic!("got ConnectResult"),
+                StreamEvent::IncomingConnectionEstablished(_) => unimplemented!(),
+                StreamEvent::ReceivedData(_, data) => {
+                    let data = String::from_utf8_lossy(&data);
+                    if &data == "\"Pong\"" {
+                        logger.log("received_pong_count", 1.0);
+                    }
+                }
+                StreamEvent::ConnectionClosed(_) => panic!("connection cloned"),
+            }
+        }
+
+        // stress test is currently just sending ping every tick
+        self.ping();
+
+        StressJobTickResult::default()
+    }
+}
+
+/// our suite creates a thread for sim2h and gives the code processor time
+struct Suite {
+    sim2h_cont: Arc<Mutex<bool>>,
+    sim2h_join: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Suite {
+    /// create a new sim2h server instance on given port
+    #[allow(clippy::mutex_atomic)]
+    pub fn new(port: u16) -> Self {
+        let sim2h_cont = Arc::new(Mutex::new(true));
+        let sim2h_cont_clone = sim2h_cont.clone();
+        let sim2h_join = Some(std::thread::spawn(move || {
+            let tls_config = TlsConfig::build_from_entropy();
+            let stream_manager = StreamManager::with_std_tcp_stream(tls_config);
+            let url = Url2::parse(&format!("wss://127.0.0.1:{}", port));
+
+            let mut sim2h = Sim2h::new(
+                Box::new(SodiumCryptoSystem::new()),
+                stream_manager,
+                Lib3hUri(url.into()),
+            );
+
+            while *sim2h_cont_clone.lock().unwrap() {
+                std::thread::sleep(std::time::Duration::from_millis(1));
+                if let Err(e) = sim2h.process() {
+                    panic!("{:?}", e);
+                }
+            }
+        }));
+
+        info!("sim2h started, attempt test self connection");
+
+        // wait 'till server is accepting connections.
+        // let this one get dropped
+        await_connection(port);
+
+        Self {
+            sim2h_cont,
+            sim2h_join,
+        }
+    }
+}
+
+impl StressSuite for Suite {
+    fn start(&mut self) {}
+
+    fn progress(&mut self, stats: &StressStats) {
+        println!("PROGRESS: {:#?}", stats);
+    }
+
+    fn stop(&mut self, stats: StressStats) {
+        *self.sim2h_cont.lock().unwrap() = false;
+        std::mem::replace(&mut self.sim2h_join, None)
```

Comment:
I have seen us use this pattern a lot, but I just learned that `self.sim2h_join.take()` accomplishes the same thing. I think.

Reply:
Oh, sweet, good to know :+1: